### ISSUE TITLE
Significant memory leak in version 2.9.0

### ISSUE THREAD

**Issue by Stephan202** [immutables/immutables#1355]
When we try to upgrade from version 2.8.8 to version 2.9.0 our medium-sized multi-module Maven build fails with _java.lang.OutOfMemoryError: Java heap space_. Based on investigation using JFR/JMC and `git bisect` the issue appears to be introduced in these two commits:
- 080040703de9ee9a6a842e783be66a3a21ff24bb (#1235).
- 2b9a79ce359d852c5c697792fd3f9f17b3c8efeb (#1249).

Some numbers:
- After the first commit meta space usage increases from ~115MB to ~260MB, while heap memory usage increases from ~270MB to ~620MB.
- After the second commit heap memory usage further increases to ~960MB.

(I performed many runs. There is little variance in these runs; on the order of ~5-10MB. After the first commit the classical GC zig-zag/sawtooth pattern all but disappears, hinting at a classical memory leak. See also the screenshots at the bottom of this message.)

JMC's _Memory -> Live Objects_ view hints at the thread local state of [`StaticEnvironment`](https://github.com/immutables/immutables/blob/15a876c0b9acd515986d6cec3fb93671dfd442d5/generator/src/org/immutables/generator/StaticEnvironment.java) being retained. Based on some subsequent debugging the following issues appear to contribute:
- `AbstractGenerator#init` calls `StaticEnvironment.preInit`, but if `AbstractGenerator#process` is not subsequently called then `StaticEnvironment#shutdown` won't be invoked.
- Although we [don't fork](https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#fork), it appears that each invocation of the `maven-compiler-plugin` loads a new `StaticEnvironment` class, which means that for each invocation a new thread-local [`EnvironmentState`](https://github.com/immutables/immutables/blob/15a876c0b9acd515986d6cec3fb93671dfd442d5/generator/src/org/immutables/generator/StaticEnvironment.java#L75-L80) is created.
- These two things combined appear to cause a significant memory leak, (I think) especially due to the accumulation of `ProcessingEnvironment`s.

I don't yet see a quick fix. Unfortunately `javax.annotation.processing.Processor` does not provide a "cleanup" or "shutdown" method. The proper fix likely involves getting rid of all this static and thread-local state (as also hinted at [here](https://github.com/immutables/immutables/blob/15a876c0b9acd515986d6cec3fb93671dfd442d5/generator/src/org/immutables/generator/StaticEnvironment.java#L29-L32)), but that seems like a large task...

That said, as-is version 2.9.0 of Immutables appears to be unusable for us. The build on which I reproduced this has 52 `pom.xml` files with 90 `maven-compiler-plugin` invocations (`compile` and `testCompile` combined), but we also have a few much-larger builds with 250+ `pom.xml` files.

---

Below three screenshots from JMC's _JVM Internals -> Garbage Collections_ view. The first build uses e31f58d44612f2211679a07229a96d9ddf62aa7a (i.e. the parent of the first problematic commit), the second uses 080040703de9ee9a6a842e783be66a3a21ff24bb, while the third uses 2b9a79ce359d852c5c697792fd3f9f17b3c8efeb. Note the _Heap Post GC_ and _Metaspace_ graphs.

![immutables-run-1](https://user-images.githubusercontent.com/513280/150645988-30945721-fc82-4c19-9466-6e733040f64b.png)
![immutables-run-2](https://user-images.githubusercontent.com/513280/150646022-bb634bdb-fa03-490e-bda0-180393b2cc3f.png)
![immutables-run-3](https://user-images.githubusercontent.com/513280/150646028-f14f8c92-365a-459c-abd5-ecb4b3ef3649.png)

**Comment by elucash** [comment]
thank you for the detailed report. I will be looking into it. On one hand it's tricky, at least I don't understand the mechanism of the but yet, on the other hand I have a gut feeling that the fix would not require significant changes. Even if it would be plausible to fix it the right way, I'm yet to formulate what is the right way is. But while we're looking at it, if anyone have precise suggestion how to fix, I would encourage to help us guessing

**Comment by elucash** [comment]
added some logging to uncover pathologies.
so far current runs give us something like this 
```
[WARNING] INIT{thread=1, thread-object=549496397, processor=330038850, processor-env=323210892}
[WARNING] STATIC-INIT{thread=1, environment-state=1764452555, thread-object=549496397, processor-env=323210892}
[WARNING] INIT{thread=1, thread-object=549496397, processor=1983900187, processor-env=1386380661}
[WARNING] STATIC-INIT{thread=1, environment-state=1992235626, thread-object=549496397, processor-env=1386380661}
[WARNING] PROCESS{thread=1, thread-object=549496397, processor=1983900187, processor-env=1386380661, round-env=650698290, processing-over=false}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=1992235626, thread-object=549496397, round-env=650698290, processor-env=1386380661}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=1992235626, thread-object=549496397, round-env=650698290, processor-env=1386380661}
[WARNING] PROCESS{thread=1, thread-object=549496397, processor=1983900187, processor-env=1386380661, round-env=925894656, processing-over=false}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=1992235626, thread-object=549496397, round-env=925894656, processor-env=1386380661}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=1992235626, thread-object=549496397, round-env=925894656, processor-env=1386380661}
[WARNING] PROCESS{thread=1, thread-object=549496397, processor=1983900187, processor-env=1386380661, round-env=16083859, processing-over=true}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=1992235626, thread-object=549496397, round-env=16083859, processor-env=1386380661}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=1992235626, thread-object=549496397, round-env=16083859, processor-env=1386380661}
```


Another run
```
[WARNING] INIT{thread=1, thread-object=1845548855, processor=1701580462, processor-env=2105513583}
[WARNING] STATIC-INIT{thread=1, environment-state=1643154089, thread-object=1845548855, processor-env=2105513583}
[WARNING] PROCESS{thread=1, thread-object=1845548855, processor=1701580462, processor-env=2105513583, round-env=1995338497, processing-over=false}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=1643154089, thread-object=1845548855, round-env=1995338497, processor-env=2105513583}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=1643154089, thread-object=1845548855, round-env=1995338497, processor-env=2105513583}
[WARNING] PROCESS{thread=1, thread-object=1845548855, processor=1701580462, processor-env=2105513583, round-env=1230539888, processing-over=false}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=1643154089, thread-object=1845548855, round-env=1230539888, processor-env=2105513583}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=1643154089, thread-object=1845548855, round-env=1230539888, processor-env=2105513583}
[WARNING] PROCESS{thread=1, thread-object=1845548855, processor=1701580462, processor-env=2105513583, round-env=1134266682, processing-over=true}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=1643154089, thread-object=1845548855, round-env=1134266682, processor-env=2105513583}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=1643154089, thread-object=1845548855, round-env=1134266682, processor-env=2105513583}
```

another sample
```
[WARNING] INIT{thread=1, thread-object=1269118128, processor=1863100050, processor-env=739264372}
[WARNING] STATIC-INIT{thread=1, environment-state=484841769, thread-object=1269118128, processor-env=739264372}
[WARNING] PROCESS{thread=1, thread-object=1269118128, processor=1863100050, processor-env=739264372, round-env=9844218, processing-over=false}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=484841769, thread-object=1269118128, round-env=9844218, processor-env=739264372}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=484841769, thread-object=1269118128, round-env=9844218, processor-env=739264372}
[WARNING] PROCESS{thread=1, thread-object=1269118128, processor=1863100050, processor-env=739264372, round-env=1936375962, processing-over=false}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=484841769, thread-object=1269118128, round-env=1936375962, processor-env=739264372}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=484841769, thread-object=1269118128, round-env=1936375962, processor-env=739264372}
[WARNING] PROCESS{thread=1, thread-object=1269118128, processor=1863100050, processor-env=739264372, round-env=931911789, processing-over=true}
[WARNING] STATIC-ROUND-INIT{thread=1, environment-state=484841769, thread-object=1269118128, round-env=931911789, processor-env=739264372}
[WARNING] STATIC-SHUTDOWN{thread=1, environment-state=484841769, thread-object=1269118128, round-env=931911789, processor-env=739264372}
```

**Comment by elucash** [comment]
here's dump of my thinking now

- definitely lacking some init - shutdown correlation
- but what contributes to the memory leak the most, accumulation of what kind of objects? (probably rounds + all cached elements etc)
- where to force cleaning and on which elements?
- maybe make environment state a part / field of the annotation processor and call proper cleanups from the abstract generator
 * only have static thread local as a way of assigning/providing current env reference, not for creation or management

**Comment by Stephan202** [comment]
@elucash thanks for jumping on this! I just ran a test against a0010e93e41402c33f45d01befc433662c385c2e and heap space ended up at ~560MB while meta space ended up at 105MB:
![immutables-after-a0010e9](https://user-images.githubusercontent.com/513280/150672747-779c0042-97ec-44e8-adc9-8256da1778b0.png)

So your changes appear to at least bring heap space usage back in line with version 2.8.8 levels. Heap usage OTOH still appears to be double of what it was before. (But: nice fix! Happy to see this didn't involve a major rewrite :sweat_smile:.)

**Comment by Stephan202** [comment]
Performed a new run based on ca0fa7ca6d1b4533815333c6ad0e4512997f1c3e, and the results look very promising:
- Meta space ended up at ~80MB (~30% reduction compared to version 2.8.8!).
- Heap space ended up at ~430MB (a ~60% increase compared to version 2.8.8).

![immutables-after-ca0fa7c](https://user-images.githubusercontent.com/513280/150727491-abf9fdeb-325c-44bf-9508-c1d41a1337c2.png)

**Comment by elucash** [comment]
I think i did my best removing as much of statics as possible while avoiding bigger redesign. Static threadlocal and accessors methods are merely "transport" mechanism now.
* It's possible that we just trading statics vs non-statics now
  - may happen hard static caches were actually preventing some duplication before
* possible improvement may come from actually analyzing heap objects to see a sort of chart of objects (counts/cumulative size) (I did smth like that with jvisualvm, but getting to do this will take day or two)
* another possibility is that throwaway context happens to capture something and dragging it along. The one with which used for init only, if that 's the case(?).

**Comment by elucash** [comment]
Created heap dumps, the picture is similar between repeated attempts on the same version, also differs between versions. I don't have an explanation, nothing strikes immediately.

<img width="1005" alt="Screen Shot 2022-01-29 at 12 04 54 AM" src="https://user-images.githubusercontent.com/1692611/151648178-c9c4fa97-703d-4854-b112-ad567b2dd86b.png">
<img width="1005" alt="Screen Shot 2022-01-29 at 12 05 04 AM" src="https://user-images.githubusercontent.com/1692611/151648180-c0ec81e4-32ca-46f5-8e53-a645c5b38849.png">

 Have those `ImportScope` etc, but this doesn't give me any idea. Thinking of just try to blindly remove context initialization on init() (sacrificing those extensions) to see if that would make any difference.

**Comment by elucash** [comment]
moving `env.initProcessing(processingEnv);` into the `processing` (guarded to only be called on the first round) did not changed the heap dump structure, just, predictably, custom immutable annotation extension stopped to work.

<img width="944" alt="Screen Shot 2022-01-29 at 1 02 49 PM" src="https://user-images.githubusercontent.com/1692611/151672120-d30c42f2-04ff-4252-8e0f-0bd65a38fb42.png">

**Comment by cal101** [comment]
Can someone share some example project for reproducing the issue? 
Our usage of immutables is not big so i do not expect major differences in use of memory but i didn't tested it yet.

**Comment by Stephan202** [comment]
@cal101 the code on which I ran the previous analyses is closed-source, unfortunately.

(@elucash sorry for not helping out more on this topic; currently very busy with other tasks. I intended to try a Jitpack'ed version of these changes against some other projects internally, but even that's still on the TODO list somewhere. Getting more feedback from others may be easier with a new 2.9.1 release (or some RC), though :thinking:.)

**Comment by Stephan202** [comment]
@elucash WDYT of something like the following patch?
```diff
diff --git a/generator/src/org/immutables/generator/EnvironmentState.java b/generator/src/org/immutables/generator/EnvironmentState.java
index 82570a12..78826b17 100644
--- a/generator/src/org/immutables/generator/EnvironmentState.java
+++ b/generator/src/org/immutables/generator/EnvironmentState.java
@@ -1,5 +1,6 @@
 package org.immutables.generator;
 
+import java.lang.ref.WeakReference;
 import java.util.Set;
 import javax.annotation.Nullable;
 import javax.annotation.processing.ProcessingEnvironment;
@@ -15,10 +16,11 @@ import com.google.common.collect.MutableClassToInstanceMap;
  * Next iteration of the same flawed design in hopes to untangle it at least a bit.
  */
 public class EnvironmentState {
-  static final ThreadLocal<EnvironmentState> currentState = new ThreadLocal<EnvironmentState>();
+  static final ThreadLocal<WeakReference<EnvironmentState>> currentState = new ThreadLocal<>();
 
   private static EnvironmentState state() {
-    return Preconditions.checkNotNull(currentState.get(), "Static environment should be initialized");
+    WeakReference<EnvironmentState> reference = Preconditions.checkNotNull(currentState.get(),"Static environment should be initialized");
+    return Preconditions.checkNotNull(reference.get(), "State should still be strongly referenced elsewhere");
   }
 
   public static <T extends Runnable> T getPerRound(Class<T> type, Supplier<T> supplier) {
@@ -60,13 +62,13 @@ public class EnvironmentState {
 
   void initProcessing(ProcessingEnvironment processing) {
     this.processing = processing;
-    currentState.set(this);
+    currentState.set(new WeakReference<>(this));
   }
 
   void initRound(Set<? extends TypeElement> annotations, RoundEnvironment round) {
     this.round = round;
     this.annotations = ImmutableSet.copyOf(annotations);
-    currentState.set(this);
+    currentState.set(new WeakReference<>(this));
   }
 
   void completeRound() {
```

IIUC the `public static` methods are only meant to be invoked between `initRound`/`initProcessing` and `completesRound`, so `reference.get()` should never be `null`. I performed a few test runs with this code, and the results are very promising: with this change the current revision of `master` performs better than version 2.8.8.

Below are three memory usage graphs against for version 2.8.8, revision e373be74921292608f668cfa69c99fc8aea32f58, and e373be74921292608f668cfa69c99fc8aea32f58 + the aforementioned patch, respectively: 

![memory-usage-2.8.8](https://user-images.githubusercontent.com/513280/163728932-a487495c-2ec3-436c-8a31-5d6adab2ba9d.png)
![memory-usage-rev-e373be74](https://user-images.githubusercontent.com/513280/163728931-bb885736-8fec-499a-8dd6-b0df9959a0ec.png)
![memory-usage-rev-e373be74-plus-patch](https://user-images.githubusercontent.com/513280/163728930-d75058c4-e018-42da-b4c2-5223816264d1.png)

(This doesn't solve the root cause of the init-shutdown correlation mismatch, but at least it's a way forward.)

**Comment by elucash** [comment]
@Stephan202 thank you! Pushed these changes. If that helps, i'm fully ok with the change. Because `EnvironmentState` is pinned as a field of the annotation processor, it should stay reachable for as long as the annotation processor instance is reachable and when the processor will be GC, the thread-local weakref will be enqueued. I still have no clear understanding how threading / annotation processing instances are managed so that we have such a leak, but if this solution works, this is great! We will be able to release new version once this fix will be confirmed plus I'll take a look at couple of new issues added recently (exception in the annotation processor + `<any>` type again etc), hopefully very soon.

**Comment by Stephan202** [comment]
Indeed, it's not fully satisfying that the root cause is unknown, but I guess that's a mystery for another day :)

Looking forward to the release! (I tested the `WeakReference` change also against a significantly larger build than the one used for the above screenshots, and there the results looked good, too.)

**Comment by Stephan202** [comment]
@elucash would it be possible to have a new release with this change? (We're interested in adopting a release that includes #1156. :) )

NB: I noticed that there are no Git tags associated to to the various `2.8.9` and `2.9.0` released found on [Maven Central](https://repo1.maven.org/maven2/org/immutables/immutables/). If they stil exist on your local machine, could you please push those?

**Comment by pingtimeout** [comment]
Hello folks.  First, thanks a lot for all the research you have been doing.  I ran into a similar situation and it was really easy to solve given that you already had a patch available.

> I still have no clear understanding how threading / annotation processing instances are managed so that we have such a leak, but if this solution works, this is great!

@elucash I may be able to provide some context here.  We are building a large Maven project composed of 450 modules.  And I noticed a pretty bad classloader leak (>5.5 GB usage in the Metaspace).

I think what is happening is that every time a module is built, a new `ClassLoader` is instantiated.  Each time, the `org.immutables:value` annotation processor is loaded in a separate class loader.  And since it stores its `StaticEnvironment` in a `ThreadLocal`, with a strong reference, the environment is never deleted.  Thus, the class definition stays reachable, and so does the class loader itself.

After ~20 modules compiled with the annotation processor, I ended up with as many leaked class loaders (including all their static variables, leading to a memory leak in the heap as well).

See this screenshot, which shows how `ThreadLocals` hold references to multiple `StaticEnvironments`, aand therefore to their class loaders.

![eclipse-mat-classloaders-path-to-gc-root](https://user-images.githubusercontent.com/1159578/185630588-2f1ab7ce-b79c-456f-a502-302b6614dee3.png)

I cloned this project and built it locally from `master`.  I can confirm that with the fix proposed by @Stephan202, the memory leak is gone.

Would it be possible to cut a release with this change please?

**Comment by elucash** [comment]
That's makes sense, thank you for more context/explanation. This change is pending the release for some time now, I just got overwhelmed with other stuff, but also I need to update gpg keys for release and this became a sort of yak shaving. I still hope to just sit and make the release asap.

**Comment by elucash** [comment]
just a heads up for 2.9.1 released, closing this, hopefully fixed :)

**Comment by pingtimeout** [comment]
Thanks !

Just one clarification, I forgot to mention that in my case, I was running with 2.8.8 when I ran into the leak, not 2.9.0.  The whole analysis is still valid as the class loader leak was caused by strong references in thread locals.  I just want to make it clear that older versions are affected as well.

So, just in case, for users who run an outdated version: upgrade to ≥2.9.1 to benefit from the WeakReference fix.
